### DISCUSSION TITLE
The future of Component Libraries for Nextjs 13.4+

### DISCUSSION THREAD

**Discussion by WillowMT** [vercel/next.js#49388] [App Router]
Hello everyone, I have recently been testing out the new NextJs app router which just went stable and must say that I love the new feature and concept it brings to us developers. It provides a ton of new features, especially the server and client components which is the highlight of this release. 

The community has done an excellent job in driving the web forward, but there was something that came to my mind when using this new server/client architecture. 

As somebody who enjoys using different UI component libraries like Bootstrap and Chakra UI, I am facing issues using my favorite UI frameworks because these libraries make use of different react hooks like context and state hooks which cannot be rendered on the server. The approach is to only use the 'use client' until the developers have updated the package to include 'use client' in every of its component. 

My question to the community would be "if a person were to only use component library through their whole nextjs app, which is client side code, wouldn't it defeat the whole purpose of having this server and client side architecture?" An example would be Chakra UI where we need a ChakraProvider at the root of the project which is a client code. Every component that comes later on would also be rendered on the client. 

I am unsure as to how we can leverage the server side to make our apps a tad faster. Typography elements such as <Heading /> could be rendered on the server instead but it would not work unless 'use client' is used so far. 

Please tell me your thoughts and ideas on this topic and the workaround if you find any. 

Thank you for taking your time reading.

Happy coding developers!

**Comment by 42shadow42** [comment]
I've been having thoughts along these same lines, with the tech as it stands you can't use hooks in a server component, and you can't use async components (safely) in a client side component. I find this lends itself to limited re-use of components which is a fundamental principle of react in my mind. I would like to see RSC and client components be consistent in functionality and by extension more re-usable. 

One thought that keeps pushing into my mind, is why it's restricted to use hooks in a server component, sure the hooks might be a bit odd in semantics in a server context but it seems no different than the hooks being usable in SSR. In testing locally in a non-nextjs setting it appears that react does in fact allow this, it does however appear that in some circumstances the hooks will fail to resolve and function, particularly when working in an async server component. If the rule became that you could use hooks in any synchronous component, I believe it would be possible reuse the components from component libraries without their adoption.

**Reply by 42shadow42** [reply]
Learning more about it, I think I found out that client components still SSR, so maybe it is a good idea to restrict some client specific hooks to client components since then they can have full functionality, the only exception I can think of which is context, which is already seeing much discussion. Whatever solution comes up for contexts I'd like it to be possible to interoperate between client/server, although this might be a huge technical challenge to overcome.
